### PR TITLE
Crd registry

### PR DESCRIPTION
**Pull request by EItanya**
rework of how solo-kit handles Crds to support multiple versions in kubernetes
BOT NOTES: 
resolves https://github.com/solo-io/solo-kit/issues/207

### CODE REVIEW

File: pkg/api/v1/clients/kube/crd/crd.go [hunk 1]
```diff
@@ -21,34 +18,49 @@ import (
...
+}
+
+type Crd struct {
+	CrdMeta
+	Version Version
+}
+
+type CombinedCrd struct {
+	CrdMeta
+	Versions []Version
```
Comment (ilackarms) [inline pkg/api/v1/clients/kube/crd/crd.go]: s/CombinedCrd/MultiVersionCrd

File: pkg/api/v1/clients/kube/crd/registry.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	if err != nil {
+		return err
+	}
+	toRegister, err := r.getKubeCrd(crd, gvk)
+	if err != nil {
+		return err
+	}
+	_, err = clientset.ApiextensionsV1beta1().CustomResourceDefinitions().Create(toRegister)
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		return fmt.Errorf("failed to register crd: %v", err)
```
Comment (rickducott) [inline pkg/api/v1/clients/kube/crd/registry.go]: extract error

File: test/mocks/api/v1/solo-kit.json [hunk 3]
```diff
@@ -3,7 +3,6 @@
   "name": "testing.solo.io",
   "version": "v1",
   "docs_dir": "../doc",
-  "crd_group_override": "crds.testing.solo.io",
```
Comment (rickducott) [inline test/mocks/api/v1/solo-kit.json]: Should add back some test for this feature

File: pkg/api/v1/clients/kube/crd/crd_test.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+					return registry.AddCrd(v)
+				})
+			}
+			Expect(eg.Wait()).NotTo(HaveOccurred())
+			Expect(registry.crds).To(HaveLen(2))
+			for _, v := range registry.crds {
+				Expect(v.Versions).To(HaveLen(10))
+			}
+		})
+		It("can retrieve avilable crds", func() {
```
Comment (rickducott) [inline pkg/api/v1/clients/kube/crd/crd_test.go]: nit: available

File: pkg/api/v1/clients/kube/crd/registry.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+	VersionExistsError = func(version string) error {
+		return errors.Errorf("tried adding version %s, but it already exists")
+	}
+
+	NotFoundError = func(id string) error {
+		return errors.Errorf("could not find the combined crd for %v", id)
+	}
+
+	InvalidGVKError = func(gvk schema.GroupVersionKind) error {
```
Comment (rickducott) [inline pkg/api/v1/clients/kube/crd/registry.go]: don't see tests for this error

File: pkg/api/v1/clients/kube/crd/registry.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/solo-io/go-utils/errors"
+	"github.com/solo-io/go-utils/kubeutils"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiexts "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type Registry struct {
```
Comment (ilackarms) [inline pkg/api/v1/clients/kube/crd/registry.go]: do these types and funcs need to be public? i don't think we should be accessing this registry anywhere outside of the CRD package, should we?

File: pkg/code-generator/codegen/templates/resource_template.go [hunk 7]
```diff
@@ -200,14 +200,27 @@ func (o *{{ .Name }}) DeepCopyObject() runtime.Object {
...
+		{{ .Name }}GVK.Group,
+		{{ .Name }}GVK.Version,
+		{{ .Name }}GVK.Kind,
+		"{{ .ShortName }}",
+		{{ .ClusterScoped }},
+		&{{ .Name }}{})
+)
+
+func init() {
+	if err := crd.GetRegistry().AddCrd({{ .Name }}Crd); err != nil {
```
Comment (ilackarms) [inline pkg/code-generator/codegen/templates/resource_template.go]: this appears to be the only place the Registry gets referenced from the outside
i think we should simply provide a package-level `crd.AddCrd` and make everything else private

**Pull request merged**